Zotero/Controllers/RecognizerController.swift

```diff
@@ -0,0 +1,585 @@
+        let authors: [Author]
+    }
+
+    struct RecognizerRequest: ApiResponseRequest {
+        typealias Response = RemoteRecognizerResponse
+        let parameters: [String: Any]?
+
+        var endpoint: ApiEndpoint { .other(URL(string: "https://services.zotero.org/recognizer/recognize")!) }
+        var httpMethod: ApiHttpMethod { .post }
+        var encoding: ApiParameterEncoding { .json }
+        var headers: [String: String]? { nil }
+
+        init(parameters: [String: Any]) {
+            self.parameters = parameters
+        }
+    }
+
+    struct RecognizerTask: Hashable {
+        enum Kind: Hashable {
+            case simple
+            case createParentForItem(libraryId: LibraryIdentifier, key: String)
+        }
+
+        let file: FileData
+        let kind: Kind
+    }
+
+    enum Error: Swift.Error {
+        case cantStartPDFWorker
+        case pdfWorkerError
+        case recognizerFailed
+        case remoteRecognizerFailed
+        case noRemainingIdentifiersForLookup
+        case unexpectedState
+        case cantCreateParentForItem
+    }
+
+    struct Update {
+        enum Kind {
+            case failed(Error)
+            case cancelled
+            case enqueued
+            case recognitionInProgress
+            case remoteRecognitionInProgress(data: [String: Any])
+            case identifierLookupInProgress(response: RemoteRecognizerResponse, identifier: String)
+            case translated(itemResponse: ItemResponse)
+            case createdParent(item: RItem)
+        }
+
+        let task: RecognizerTask
+        let kind: Kind
+    }
+
+    enum RecognizerTaskState {
+        case enqueued
+        case recognitionInProgress
+        case remoteRecognitionInProgress(data: [String: Any])
+        case identifiersLookupInProgress(response: RemoteRecognizerResponse, currentIdentifier: RemoteRecognizerIdentifier, pendingIdentifiers: [RemoteRecognizerIdentifier])
+    }
+
+    // MARK: Properties
+    private unowned let pdfWorkerController: PDFWorkerController
+    private unowned let apiClient: ApiClient
+    private unowned let translatorsController: TranslatorsAndStylesController
+    private unowned let schemaController: SchemaController
+    private unowned let dbStorage: DbStorage
+    private unowned let dateParser: DateParser
+    private let dispatchSpecificKey: DispatchSpecificKey<String>
+    private let accessQueueLabel: String
+    private let accessQueue: DispatchQueue
+    private let backgroundQueue: DispatchQueue
+    private let updatesSubject: PublishSubject<Update>
+    var updates: Observable<Update> {
+        updatesSubject.asObservable()
+    }
+    private let disposeBag: DisposeBag
+
+    internal weak var webViewProvider: WebViewProvider?
+
+    // Accessed only via accessQueue
+    private static let maxConcurrentRecognizerTasks: Int = 1
+    private var queue: OrderedDictionary<RecognizerTask, (state: RecognizerTaskState, observable: PublishSubject<Update>)> = [:]
```

Comment:
Maybe consider having separate "wait queue" and "in progress dictionary", same as above. But it's not as apparent here, so we might as well just keep this.